### PR TITLE
Build a MLP

### LINKED ISSUE #12: Add Leaky ReLu

### PR DESCRIPTION
**Pull request by rlouf**
This PR adds enough functionality to build a MLP with AeNN: a dense layer, ReLu and Leaky ReLu activation layers.

This should be enough to demonstrate what we can do with rewrites in Aesara:

- [x] Replace layers
- [x] Replace activation functions
- [x] Add new layers
- [ ] Turn a NN into a BNN (replace variable with random variable)

Bayesian neural networks, of course:
- [ ] Use AePPL to compute logdensity for NN with normal priors

Of course we should be able to compile this to different targets:
- [ ] C
- [ ] Numba
- [ ] JAX

And train it:
- [ ] Optimiser
- [ ] Loss function
- [ ] State initialisation

### CODE REVIEW

**Comment by rlouf** [comment]
It is fairly easy to define layer types using `OpFromGraph`; to make the library more user-friendly we can probably take care of the boilerplate code necessary to create an `OpFromGraph` in a parent class.

We can already swap activation layers with rewrites, however trying to change the layers' size will fail because the outputs of the replaced and replacing nodes are not of the same type. ~~Is that a fundamental limitation of the rewrites? Can we bypass it?~~ The size information should be stored as an `Op` property and we should initialise the weights without a specified shape; size information is only necessary when initialising the weights.

Regarding the rewrites, a few other remarks:
1. "Apply on all (one pass)" should have an easier interface
2. "Apply on the nth occurence of the pattern" should have an interface
3. "Apply on one occurence at random" should have an interface

All of this can be implemented here, wrapping Aesara's rewrite machinery.1

**Comment by rlouf** [comment]
The state management is going to be fun; we should find a system to automatically name the layers, like Flax does.

Also, it would be easier to use names to target objects with  rewrites so we should allow naming.

Rewrites could do with a wrapper function.

The `DenseLayer` class is not doing any work, we only really use the `__call__` method. Consider using closures instead.

**Comment by rlouf** [comment]
Regarding BNNs, we can "expand" random variables by default to an array of iid variables to have the following API:

```python
import aesara.tensor as at

srng = at.random.RandomStream(0)

x = at.vector('x')

w = srng.normal(0, 1)
y_rv = Dense(50, w)(x)
```
